rfc/20240529-OpenTofu-Specific-Code-Override.md

```diff
@@ -0,0 +1,160 @@
+3. In the future, Terraform might introduce new features that are not supported by OpenTofu.
+
+This divergence between OpenTofu and Terraform configuration syntax imposes challenges on the following personas:
+1. Module maintainers who want their modules to be compatible with both OpenTofu and Terraform.
+2. Users who want to use OpenTofu, but are missing IDE and 3rd party tools support for OpenTofu configuration files.
+3. 3rd party tools maintainers who want to support both OpenTofu and Terraform.
+
+## Proposed Solution
+
+In addition to supporting `.tf` extension files as we do today, OpenTofu will support `.tofu` extension files. When two files have the same name, one with the `.tf` extension and one with the `.tofu` extension, OpenTofu will load only the `.tofu` file and ignore the `.tf` file. Hence, creating a mechanism that overrides specific files. This way the `.tofu` extension files will be used for OpenTofu specific features and syntax and the `.tf` extension files can be used for Terraform compatible features and syntax. Allowing users and module maintainers to use OpenTofu and Terraform in the same project without conflicts. 
+
+For example, If I have two files:
+```hcl
+// vars.tf
+
+variable "region" {
+  type = string
+}
+```
+
+```hcl
+// vars.tf
+
+variable "region" {
+  type = string
+   deprecated = "Reason this was deprecated"
+}
+```
+Terraform would read all files including `vars.tf` and ignore `vars.otf`.
+OpenTofu would read all files, but replace `vars.tf` with `vars.otf` and will include the additional deprecated attribute.
+
+
+> [!NOTE]
+> This solution should not be confused with the [Override Files feature](https://opentofu.org/docs/language/files/override/) that is already supported in OpenTofu. The Override Files feature is used to override specific portions of an existing file, while the proposed solution here will override the **entire file**. We chose to implement it this way because Terraform may introduce features in the future that OpenTofu cannot parse, making specific overrides insufficient.
+
+## User Documentation
+
+### Supported Scenarios
+As we know, `.tf` are not the only files that OpenTofu can load, in here we will drill down into all the supported scenarios:
+1. this is the basic scenario we mentioned before, if a `.tofu` file exists, we’ll load it instead of the `.tf` file.
+2. For [JSON Configuration](https://opentofu.org/docs/language/syntax/json/), if a `.tofu.json` file exists, we’ll load it instead of the `tf.json` file. 
+3. For [test files](https://opentofu.org/docs/cli/commands/test/), if `.tofutest.hcl`/`.tofutest.json` files exist, we’ll load them instead of the `.tftest.hcl`/`.tftest.json` files. 
+4. For [override files](https://opentofu.org/docs/language/files/override/), if `_override.tofu`/`_override.tofu.json` files exist, we’ll load them instead of the `_override.tf`/`_override.tf.json` files.
```

Comment:
Do we want to include `.tofuvars` (`.tfvars`) in this change or is that out of scope.  I'm not sure if that format will need to change, but it is worth considering.

Reply:
I forgot to write about it as part of this RFC. I thought about it, but as you say, I can't see how the formatting there will change, so I didn't include it. But maybe that's confusing, and users might expect us to support overriding `.tfvars` with `.tofuvars`?
Now I'm unsure, and I'll add it to the open questions sections.

Reply:
https://github.com/opentofu/opentofu/pull/1699/commits/e0e03f6ccb94b6509111cb7f8984d793c69b4709